### PR TITLE
[DPE-5053] Add spark history server monitoring

### PR DESCRIPTION
**Pull request by welpaolo**
This PR add monitoring to the Spark History Server charm

### CODE REVIEW

File: metadata.yaml [hunk 1]
```diff
@@ -20,7 +20,7 @@ resources:
   spark-history-server-image:
     type: oci-image
     description: OCI image for Spark History Server
-    upstream-source: ghcr.io/canonical/charmed-spark@sha256:4c4e6f9d394348a26ec66969898434f007467b2caee03b2d393ddffd14dc2ecf # spark-version: 3.4.2 revision: 11ea444ff878f57f92c84190a6a1a2dedadd1f3f 
+    upstream-source: ghcr.io/welpaolo/charmed-spark@sha256:90c3b889a1af438661afa431993e262503ca4ed8f4993c35149248c8e3107af3
```
Comment (theoctober19th) [inline metadata.yaml:23]: Do we have this image published upstream as well?

File: tests/integration/test_charm_logs.py [hunk 2]
```diff
@@ -249,4 +259,117 @@ async def test_loki_integration(ops_test: OpsTest, charm_versions):
             c = c + 1
     logger.info(f"Number of line found: {c}")
     assert c > 0
+
+
+@pytest.mark.abort_on_fail
+async def test_history_server_cos_relation_joined(ops_test: OpsTest, charm_versions):
+    # Prometheus data is being published by the app
```
Comment (theoctober19th) [inline tests/integration/test_charm_logs.py:266]: chore: add docstrings

**Review by theoctober19th — APPROVED** [review]
LGTM

File: metadata.yaml (hunk 1 above)
Comment (welpaolo) [reply metadata.yaml:23]: Working on it right now

Comment (welpaolo) [reply metadata.yaml]: Done

File: .github/workflows/ci.yaml [hunk 3]
```diff
@@ -93,7 +93,7 @@ jobs:
           echo "IPADDR=$(ip -4 -j route get 2.2.2.2 | jq -r '.[] | .prefsrc')" >> $GITHUB_ENV
       - name: Setup operator environment
         # TODO: Replace with custom image on self-hosted runner
-        uses: charmed-kubernetes/actions-operator@main
+        uses: charmed-kubernetes/actions-operator@5193238c86d58e874ad3457da2df640ef138a41f
```
Comment (deusebio) [inline .github/workflows/ci.yaml:96]: **todo** this can be reverted

File: pyproject.toml [hunk 4]
```diff
@@ -35,6 +35,14 @@ pydantic ="^1.10.7"
...
 boto3 = ">=1.34.55"
 jsonschema = "^4.19.1"
+# FIXME: Unpin once rustc 1.76 is available at build time
+rpds-py = "<0.19"
+
+[tool.poetry.group.charm-libs.dependencies]
+ops = "^2.13.0"
+poetry-core = "^1.9.0"
+# grafana_agent/v0/cos_agent.py requires pydantic <2
+pydantic = "^1.10.17, <2"
```
Comment (deusebio) [inline pyproject.toml:45]: **praise** this is very good and very needed! Thanks!

**Review by deusebio — APPROVED** [review]
Great stuff @welpaolo ! Thanks for also adding both dahsboards and alerts!

**Pull request merged**